File: app/src/main/org/runnerup/features/DetailSyncController.java
package org.runnerup.features;

import android.content.Intent;
import android.net.Uri;
import android.text.TextUtils;
import android.view.View;
import android.widget.CompoundButton;
import android.widget.CompoundButton.OnCheckedChangeListener;
import androidx.appcompat.app.AlertDialog;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import org.runnerup.sync.SyncManager;

/** Upload/sync account state for {@link DetailActivity}. */
final class DetailSyncController {

  final HashSet<String> pendingSynchronizers = new HashSet<>();
  final HashSet<String> alreadySynched = new HashSet<>();
  final Map<String, String> synchedExternalId = new HashMap<>();

  void openAccountUrl(DetailActivity activity, SyncManager syncManager, String name) {
    if (synchedExternalId.containsKey(name) && !TextUtils.isEmpty(synchedExternalId.get(name))) {
      String url =
          syncManager.getSynchronizerByName(name).getActivityUrl(synchedExternalId.get(name));
      if (url != null) {
        activity.startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(url)));
      }
    }
  }

  View.OnLongClickListener createClearUploadListener(
      DetailActivity activity, SyncManager syncManager, long activityId, Runnable onCleared) {
    return arg0 -> {
      final String name = (String) arg0.getTag();
      new AlertDialog.Builder(activity)
          .setTitle("Clear upload for " + name)
          .setMessage(org.runnerup.common.R.string.Are_you_sure)
          .setPositiveButton(
              org.runnerup.common.R.string.Yes,
              (dialog, which) -> {
                dialog.dismiss();
                syncManager.clearUpload(name, activityId);
                onCleared.run();
              })
          .setNegativeButton(
              org.runnerup.common.R.string.No, (dialog, which) -> dialog.dismiss())
          .show();
      return false;
    };
  }

  View.OnClickListener createUploadListener(
      SyncManager syncManager,
      long activityId,
      UploadingState uploading,
      Runnable onComplete) {
    return v -> {
      uploading.set(true);
      syncManager.startUploading(
          (synchronizerName, status) -> {
            uploading.set(false);
            onComplete.run();
          },
          pendingSynchronizers,
          activityId);
    };
  }

  OnCheckedChangeListener createSendCheckedListener(Runnable updateUploadVisibility) {
    return new OnCheckedChangeListener() {
      @Override
      public void onCheckedChanged(CompoundButton arg0, boolean arg1) {
        final String name = (String) arg0.getTag();
        if (alreadySynched.contains(name)) {
          arg0.setChecked(true);
        } else {
          if (arg1) {
            pendingSynchronizers.add(name);
          } else {
            pendingSynchronizers.remove(name);
          }
          updateUploadVisibility.run();
        }
      }
    };
  }

  interface UploadingState {
    void set(boolean uploading);
  }
}

File: app/src/main/kotlin/org/runnerup/features/DetailSyncController.kt
/*
 * Copyright (C) 2012 - 2013 jonas.oreland@gmail.com
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 */

package org.runnerup.features

import android.content.Intent
import android.net.Uri
import android.text.TextUtils
import android.view.View
import android.widget.CompoundButton
import androidx.appcompat.app.AlertDialog
import org.runnerup.sync.SyncManager

/** Upload/sync account state for [DetailActivity]. */
internal class DetailSyncController {

  @JvmField val pendingSynchronizers = HashSet<String>()
  @JvmField val alreadySynched = HashSet<String>()
  @JvmField val synchedExternalId = HashMap<String, String>()

  fun openAccountUrl(activity: DetailActivity, syncManager: SyncManager, name: String) {
    if (synchedExternalId.containsKey(name) && !TextUtils.isEmpty(synchedExternalId[name])) {
      val url = syncManager.getSynchronizerByName(name).getActivityUrl(synchedExternalId[name])
      if (url != null) {
        activity.startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(url)))
      }
    }
  }

  fun createClearUploadListener(
      activity: DetailActivity,
      syncManager: SyncManager,
      activityId: Long,
      onCleared: Runnable,
  ): View.OnLongClickListener =
      View.OnLongClickListener { view ->
        val name = view.tag as String
        AlertDialog.Builder(activity)
            .setTitle("Clear upload for $name")
            .setMessage(org.runnerup.common.R.string.Are_you_sure)
            .setPositiveButton(org.runnerup.common.R.string.Yes) { dialog, _ ->
              dialog.dismiss()
              syncManager.clearUpload(name, activityId)
              onCleared.run()
            }
            .setNegativeButton(org.runnerup.common.R.string.No) { dialog, _ -> dialog.dismiss() }
            .show()
        false
      }

  fun createUploadListener(
      syncManager: SyncManager,
      activityId: Long,
      uploading: UploadingState,
      onComplete: Runnable,
  ): View.OnClickListener =
      View.OnClickListener {
        uploading.set(true)
        syncManager.startUploading(
            { _, _ ->
              uploading.set(false)
              onComplete.run()
            },
            pendingSynchronizers,
            activityId,
        )
      }

  fun createSendCheckedListener(updateUploadVisibility: Runnable): CompoundButton.OnCheckedChangeListener =
      CompoundButton.OnCheckedChangeListener { button, checked ->
        val name = button.tag as String
        if (alreadySynched.contains(name)) {
          button.isChecked = true
        } else {
          if (checked) {
            pendingSynchronizers.add(name)
          } else {
            pendingSynchronizers.remove(name)
          }
          updateUploadVisibility.run()
        }
      }

  fun interface UploadingState {
    fun set(uploading: Boolean)
  }
}
